feat(client): display FSD in multi-level accordion

huyenltnguyen opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

This PR:
- Requires #56795 for the super block data
- Requires #56808 for the block layout split. FSD is currently sharing the component with other superblocks, so the split allows us to make styling adjustments without interfering with those
- Updates the FSD super page to display the curriculum in a multi-level accordion 
  - I'm not using the term "tree view" because of a change in the implementation approach (see the "Implementation" section below). The display is not affected by this change, though.
- Is related to https://github.com/freeCodeCamp/fCC10/issues/6

This is a _very_ rough version. Feedback and further iterations are needed. A couple things I think we could do:
- Make chapter labels display similarly to super blocks (with a border and an icon)
  - Icon: I think we could give HTML, CSS, JS, etc. icons to the corresponding chapter
  - Color: I'll need some help here if we want to apply a border and background colors to the chapters. I tried making them look like super block items, but IMHO they don't blend very well with the nested content
    <details>
      <summary>Screenshot</summary>

    <img width="776" alt="Screenshot 2024-10-28 at 23 31 31" src="https://github.com/user-attachments/assets/6f0e9f97-f2e2-4419-9aed-b64611b4717b">
      </details>

## Screenshots
<details>

<summary>Screenshots of the changes</summary>

| Level |  Screenshot |
| --- | --- |
| Chapters | <img width="788" alt="Screenshot 2024-11-01 at 03 54 50" src="https://github.com/user-attachments/assets/b157ec2f-b239-4ddd-b272-18ba2c372986"> |
| Modules | <img width="795" alt="Screenshot 2024-10-29 at 02 05 03" src="https://github.com/user-attachments/assets/84e7a829-a166-4f62-a129-4c2b699d82cb"> |
| Blocks | <img width="778" alt="Screenshot 2024-10-29 at 02 05 48" src="https://github.com/user-attachments/assets/fccb8b0a-cd51-47ec-8333-ee0607569579"> |

</details>

## Implementation approach

### TL;DR

I experimented with the tree view pattern using Material UI, got the functionality we want, but came across a performance issue.

I also wasn't happy with the keyboard navigation pattern, so I explored a little more and fell into an a11y rabbit hole. Basically the behavior we want to achieve with tree view is quite complicated, and the specification is still being discussed by the Internet Elders.

I decided to just display the content as a nested accordion. The UI is identical to my previous version, but the a11y is better (or at least less awkward).

### Details

<details>

<summary>Implementation approach in detail</summary>

Before we get to the details, here is an overly simplified definition of "tree view pattern":
- The pattern is used to organized elements in a hierarchical list.
- In this pattern, the outer-most element (or root node) has the `tree` role, and all of its children (child nodes) have the `treeitem` role regardless of whether or not they have child nodes themselves; the grandchild nodes would also have the `treeitem` role.
- In this pattern, users will navigate the tree using Arrow keys, instead of Tab and Enter/Spacebar like with most interactive elements:
  - Left or Right Arrow keys to expand or collapse a tree item.
  - Up or Down Arrow keys to move up or down the node level.
- The `treeitem` role only accepts non-interactive elements.
  - This is not entirely true, or at least the HTML specification is currently unclear. But even if interactive elements are allowed, it is still not clear how the keyboard interaction would be.
  - See https://github.com/w3c/aria/issues/1251 and https://github.com/w3c/aria/issues/1440.

The keyboard interaction is what I'm concerned about. Our `ChallengeList` and `ChallengeGrid` blocks render the challenge list within a widget (a panel that is only visible when its trigger button is clicked), so if we put them in a `tree` we have the following DOM:

```html
<!-- Pseudo code -->
<ul role="tree">
  <li role="treeitem">Chapter One</li>
    <li role="treeitem">Module One</li>	
      <li role="treeitem">
        <button onClick={expandChallengeList}>Block One</button>
        <ul>
          <li>Challenge One</li>
          <li>Challenge Two</li>
        </ul>
      </li>
    </li>
  </li>	
</ul>
```

The keyboard interaction of this structure would be cumbersome, since it involves a mix of Arrow keys (to go down to the last `treeitem`), Tab (to focus on the button), and Enter/Space key (to trigger the button). And since this is a scenario that is being discussed in the W3C issues above, I'm not comfortable implementing it without knowing the a11y requirements/recommendations.

I decided to render the curriculum as a multi-level accordion instead. The hierarchy is communicated by the `ul` and `li`  elements under the hood, and visually by the indentation.

### Other implementation notes

(This is just for posterity. I'm not going with tree view so Material UI isn't needed.)

The rendering seems to be fine, I didn't experience any lag in the display. However, the clean up operation broke the page. 

The reason is that, on unmount, MUI (the library we use for Tree View) calls a [`getNodesToRemove` function](https://github.com/mui/material-ui/blob/85e421035df0d7b22f122f8f8d65a120c38436e3/packages/material-ui-lab/src/TreeView/TreeView.js#L408-L421). The function runs recursively down the tree to get all of the tree nodes, including the nodes' content, meaning it would run against all challenges listed in the blocks. And I think the thousands of function calls trigger the error.

To reproduce:
- On /learn, go to the FSD super block
- On the FSD super block page, click on another super block
- See that the page will throw "too much recursion"

(Non-)Solution:

I wasn't sure what could be done on our end to resolve the issue. Due to the React version constraint, we can only use a deprecated MUI version, so we aren't really in a position to file a bug report / support ticket.

I also could not find a suitable alternative library (there are multiple tree view libraries, though they either are accessible but don't support our use cases, or they do support our use cases but are not accessible).

The decision to use accordion pattern was also partially because of this.

</details>

## Notes

Aside from further cosmetic improvements, we will need to:
- Handle the disabled state in chapter level, module level, and block level
- Display the completion check mark next to chapter and module title

I'll create separate PRs for those to keep this PR in a reviewable size.

<!-- Feel free to add any additional description of changes below this line -->

huyenltnguyen committed:
fix: rename component

huyenltnguyen committed:
fix: typecheck [skip ci]

huyenltnguyen committed:
fix: add accessible name to the tree

huyenltnguyen committed:
fix: memoize the tree data [skip ci]

huyenltnguyen committed:
refactor: rendering logic [skip ci]

jdwilkin4 commented:
Hey @huyenltnguyen ! 

This looks really cool so far. 🥳 

My only concern with the design is that I am not sure we have solved the issue of campers starting wherever they want to.
In the past, any time we have a list of blocks, we always get questions of should they start at the top or can they just pick randomly from the list.

For example:
https://forum.freecodecamp.org/t/where-to-start-with-fcc/591803

So, is there any way to make it super clear in the chapters list that you can only start at the top? Maybe the other content can be greyed out or someway indicated that it is not available and you have no choice but to start at the top?

jdwilkin4 commented on client/i18n/locales/english/intro.json:
```diff
@@ -1695,6 +1695,72 @@
   "front-end-development": {
     "title": "Front End Development",
     "intro": ["placeholder"],
+    "chapters": {
```
Do we want the welcome block from Quincy to be it's own chapter?
Right now it is listed as a module, but maybe it works for it to be its own chapter 🤔

jdwilkin4 replied:
I think intro works for the title and the structure looks good to me 👍

moT01 replied:
I added that [chapter here](https://github.com/freeCodeCamp/freeCodeCamp/pull/57151/files#diff-d5befb16f2e5667940f0cd15dba286e6a04c96343d8f6d2828e4f6332f0807a0R3) - I just called it `freecodecamp`, but I'm not sure that's the best.

moT01 commented:
> My only concern with the design is that I am not sure we have solved the issue of campers starting wherever they want to.

I'm not sure that needs to be solved on this PR. We do intend to add something to let people know that they skipped some challenges.

huyenltnguyen committed:
fix: convert the tree to accordion (and why is a11y so hard?)

huyenltnguyen committed:
feat: auto-expand chapter and module based on url hash [skip ci]

huyenltnguyen committed:
fix: remove default ul padding [skip ci]

huyenltnguyen committed:
feat: add MUI libs

huyenltnguyen committed:
feat: implement tree view

huyenltnguyen committed:
fix: render Block as TreeItem label

huyenltnguyen committed:
fix: manually render the tree items

huyenltnguyen committed:
feat: add chapter and module titles and styling

huyenltnguyen committed:
WIP feat: try using react-accessible-treeview

huyenltnguyen committed:
fix: cert name and strings

huyenltnguyen committed:
fix: display exam

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into feat/fed-tree

huyenltnguyen committed:
fix: alignment and icon size

huyenltnguyen committed:
fix: block bottom margin

huyenltnguyen committed:
feat: add freeCodeCamp chapter to intro.json

huyenltnguyen committed:
Revert "fix: block bottom margin"

This reverts commit 0c30d45f84a24494d30b2e70fef3304d54abce03.

huyenltnguyen committed:
Update client/src/templates/Introduction/components/block.tsx

Nirajn2311 reviewed (commented):
All good from mobile app perspective

ojeytonwilliams commented on client/src/templates/Introduction/components/super-block-accordion.tsx:
```diff
@@ -0,0 +1,150 @@
+  superBlock,
+  chosenBlock
+}: SuperBlockTreeViewProps) => {
+  const content = useMemo(() => {
```
`useMemo` can help here, but currently it's not doing a lot. While `challenges` never changes, `chosenBlock` does and each time it does we recalculate a few things we don't need to.

So, rather than putting the whole thing in a memo, we can memoize just the parts that depend on challenges.

I'll push a commit in a bit so you can see what I'm getting at.

ojeytonwilliams committed:
refactor: memoize challenges

Otherwise React cannot tell that challenges have not actually changed
since the map/filter creates new arrays each time.

ojeytonwilliams committed:
refactor: reorganise chapter/module/block data

ojeytonwilliams committed:
fix: block name of exam block

ojeytonwilliams committed:
refactor: remove unnecessary returns

ojeytonwilliams commented:
Hey @huyenltnguyen, I ended up hacking at the superblock accordion for a bit and I think it's a little clearer now.

The basic ideas are as follows 

* do the (potentially) expensive stuff in the memo, but leave the majority of code outside. 
* transform the flat challenge structure into chapters, modules and blocks to make rendering simpler

That's it, really! Let me know what you think and I'll finish the review.

By the way, I love the design and how easy it is to navigate with the keyboard!

ojeytonwilliams reviewed (approved):
LGTM :+1:

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into feat/fed-tree

moT01 reviewed (approved):
This is awesome @huyenltnguyen 🎉 🎉 🎉 

I see a number of things that can be improved on, but this looks and functions good - we can iterate on it.
